lib/spack/spack/package.py

```diff
@@ -1129,6 +1137,33 @@ def do_patch(self):
         else:
             touch(no_patches_file)
 
+    def patches_to_apply(self):
+        """If the patch set does not change between two invocations of spack,
+           then all patches in the set will be applied in the same order"""
+        patchesToApply = itertools.chain.from_iterable(
+            patch_list
+            for spec, patch_list in self.patches.iteritems()
+            if self.spec.satisfies(spec))
+        return sorted(patchesToApply, key=lambda p: p.path_or_url)
+
+    def content_hash(self, content=None):
+        """Create a hash based on the sources and logic used to build the
+        package. This includes the contents of all applied patches and the
+        contents of applicable functions in the package subclass."""
+        hashContent = list()
+        source_id = fs.for_package_version(self, self.version).source_id()
+        if not source_id:
+            message = 'Missing a source id for {s.name}@{s.version}'
+            tty.warn(message.format(s=self))
+            hashContent.append('')
```

Comment:
@mathstuf: I'd still leave a `TODO` here (in addition to the warning) with a note to fix this when we refactor fetchers.

I think you said in an earlier comment that this was brittle, and that if a tarball is removed, a package would then have a different hash.  Can you elaborate on that?  The hash for an existing *installation* will never change, because we store the hash that was generated at install time.  So this may cause a rebuild if, e.g., the version is later checksummed, but it shouldn't cause any existing install to change.

Reply:
Now that all sources and patches come with a `.sha256`, that's not a problem I think. I've put the comment back with the updated reasoning.

Reply:
thanks!